docs/rfcs/1005-realm-key-rotation.md

```diff
@@ -0,0 +1,328 @@
+The rational for this approach was that the client only have to deal with a single key at any given time.
+However this has some downsides:
+
+- Reencrypting all vlobs takes some time (though not that much in pratice given it's only
+  metadata, the empirical rule being 1000 times less metadata than actual data in Parsec).
+- This concept must be exposed to the user so that he can decide when to do the reencryption
+  maintenance procedure. This is a concept complex to explain so not great for UX :/
+- A realm encryption revision index is kept server-side and send to the client to ensure
+  it is not using the wrong key to decrypt
+- A re-encryption API is needed (the `realm_start|stop_reencryption_maintenance` &
+  `vlob_maintenance_get|save_reencryption_batch` commands)
+- The concept of maintenance has been introduced: a realm under maintenance cannot be read/write
+- Due to the maintenance, a reencryption must finish no matter what. So all realm owner have to
+  watch for ongoing maintenance an participate in it (even if they are not the initiator of it),
+  just in case the initiator disconnects (to be never seem again) before completion.
+
+On top of that the added security is small (remember, this was a practical choice) given
+the client could have downloaded all metadata when it has access (typically when doing
+file content a search on the workspace mountpoint).
+In addition since Parsec 2.16 the manifests are eagerly downloaded (instead of the
+previous lazy approach), so the reencryption really has no security benefit.
+
+And now the final blow: the maintenance preventing read/write access was such a strong
+limitation that we ended up adding a feature to keep read access possible.
+Now what that means is the client has to in fact deal with up to *two* keys:
+
+1. when a maintenance starts, the client receives the new key
+2. it overwrites the realm key with the new one it the user manifest
+3. now it tries to access some realm data
+   - the maintenance is done, so the new key works. All is good.
+   - the maintenance is still ongoing, so the client has to look into the previous
+     versions of the user manifest to find the old key. This is messy :'(
+
+## General approach
+
+The idea is as follow:
+
+- remove the whole maintenance logic: a realm is always accessible in read&write
+- remove the whole re-encryption logic: realm vlobs are never reencrypted
+- key rotation consist of broadcasting the new key in a message to each realm participant
+- key rotation also involve the creation of a certificate, this simplifies message
+  validation and allow auditability (know when and by who the key rotation occured)
+- each key for a given realm has an index (enforced by the server to avoid
+  concurrent creation issues)
+- vlob create/update/read API commands contains key index information
+- user manifest store all keys ordered by index for each realm
+- sharing a realm consist of sending a message containing all keys ordered by index
```

Comment:
Does the only up-most recent one is used for decryption/encryption (read&write) whereas the other ones are used only for decryption (read) ?

Reply:
> Does the only up-most recent one is used for decryption/encryption (read&write) whereas the other ones are used only for decryption (read) ?

yup ;-)

> If that the case, then the old keys will become garbage data (not used for encryption and all the current data use most recent keys).

This only occurs once all the old data have been replaced, which is unlikely: a file is often worked on for some time, then become effectively read-only (i.e. the document is finished and now stays there as reference). Think for instance of the `LICENSE` in a Github repo.

On top of that, old keys are also needed to access the history.

>  Should we consider system to remove useless keys ? (I currently think that it may cause a problem with clients that have a high-rate of renewal of users after some time)

A key is 32bytes long, so a 1mo keys-bundle would contain ~32000 keys, enough to do 10 keys rotation per day during 10 years.
On top of that, we can do un-sharing can be done in batch to only do a single key-rotation.

I would say the issue is more on the data size end: 1mo key bundle and 10 keys rotation per year means 3.5Go of keys-bundle data at the end of the year ! But 1) that some extreme values and 2) it is possible for the server to remove old keys-bundles (old keys-bundles are only used to do self-heal if the last key bundle is corrupted, so it is enough to only keep e.g. the last month of keys-bundles)